lean4/src/putnam_2023_a4.lean

```diff
@@ -0,0 +1,19 @@
+import Mathlib
+
+/--
+Let $v_1, \ldots, v_{12}$ be unit vectors in $\mathbb{R}^3$ from the origin to the vertices of a regular icosahedron. Show that for every vector $v \in \mathbb{R}^3$ and every $\epsilon > 0$, there exist integers $a_1, \ldots, a_{12}$ such that $\|a_1v_1 + \cdots + a_{12}v_{12} - v\| < ε$.
+-/
+theorem putnam_2023_a4
+    (v : Fin 12 → EuclideanSpace ℝ (Fin 3))
+    (hv :
+      letI α : ℝ := (1 + √5) / 4
+      letI h : ℝ := 1 / 2
+      letI e : (Fin 3 → ℝ) ≃ EuclideanSpace ℝ (Fin 3) := (WithLp.equiv _ _).symm
+      ∃ g : EuclideanSpace ℝ (Fin 3) ≃ₗᵢ[ℝ] EuclideanSpace ℝ (Fin 3),
+      g ∘ v = e ∘ ![![h, α, 0], ![-h,  α,  0], ![ h, -α,  0], ![-h, -α,  0],
```

Comment:
These strictly aren't well-typed; the inner `![`s should be elements of euclidean space, but here they are raw vectors.

leanprover-community/mathlib4#17732 will give a path to fixing this eventually, but in the meantime you should probably add the missing `(WithLp.equiv _ _).symm` before each one.

Reply:
The vectors given by (\pm h, \pm a, 0) or any cyclic perm are not actually unit here. Per the solution it doesn't matter but it skips a reasoning step to use this form. (Thanks to Michael for spotting this!)

Reply:
@eric-wieser

Surely the typing is correct. Indeed `e` has type `(Fin 3 → ℝ) ≃ EuclideanSpace ℝ (Fin 3)` and
```lean
  ![![h, α, 0], ![-h,  α,  0], ![ h, -α,  0], ![-h, -α,  0],
    ![α, 0, h], ![ α,  0, -h], ![-α,  0,  h], ![-α,  0, -h],
    ![0, h, α], ![ 0, -h,  α], ![ 0,  h, -α], ![ 0, -h, -α]]
```
has type `Fin 12 → Fin 3 → ℝ` so after composing with `⇑e` we get a map `Fin 12 → EuclideanSpace ℝ (Fin 3)`.

Reply:
> The vectors given by (\pm h, \pm a, 0) or any cyclic perm are not actually unit here. 

Thanks. Urgh, it looks like I fixed the edge lengths at 1 instead of placing the vertices on the unit sphere. I've (force) pushed a fix for this.

Reply:
Ah apologies, I missed `e` being here